src/app/clusters/ota-requestor/OTARequestor.cpp

```diff
@@ -252,36 +244,39 @@ EmberAfStatus OTARequestor::HandleAnnounceOTAProvider(app::CommandHandler * comm
     }
     ChipLogDetail(SoftwareUpdate, "  Endpoint: %" PRIu16, providerLocation.endpoint);
 
-    // If reason is URGENT_UPDATE_AVAILABLE, we start OTA immediately. Otherwise, respect the timer value set in mOtaStartDelayMs.
-    // This is done to exemplify what a real-world OTA Requestor might do while also being configurable enough to use as a test app.
-    uint32_t msToStart = 0;
-    switch (announcementReason)
-    {
-    case OTAAnnouncementReason::kSimpleAnnouncement:
-    case OTAAnnouncementReason::kUpdateAvailable:
-        msToStart = mOtaStartDelayMs;
-        break;
-    case OTAAnnouncementReason::kUrgentUpdateAvailable:
-        msToStart = kImmediateStartDelayMs;
-        break;
-    default:
-        ChipLogError(SoftwareUpdate, "Unexpected announcementReason: %u", static_cast<uint8_t>(announcementReason));
-        return EMBER_ZCL_STATUS_FAILURE;
-    }
-
-    DeviceLayer::SystemLayer().StartTimer(System::Clock::Milliseconds32(msToStart), StartDelayTimerHandler, this);
+    mOtaRequestorDriver->ProcessAnnounceOTAProviders(providerLocation, announcementReason);
 
     return EMBER_ZCL_STATUS_SUCCESS;
 }
 
 void OTARequestor::ConnectToProvider(OnConnectedAction onConnectedAction)
 {
-    VerifyOrReturn(mOtaRequestorDriver != nullptr, ChipLogError(SoftwareUpdate, "OTA requestor driver not set"));
-    VerifyOrReturn(mServer != nullptr, ChipLogError(SoftwareUpdate, "Server not set"));
-    VerifyOrReturn(mProviderLocation.HasValue(), ChipLogError(SoftwareUpdate, "Provider location not set"));
+    // We are now connecting to a provider, leave the kIdle state.
+    // No state matches this one fully but we can't be in kIdle.
+    RecordNewUpdateState(OTAUpdateStateEnum::kUnknown, OTAChangeReasonEnum::kSuccess);
```

Comment:
Why can't we be in idle? We technically haven't even begun querying yet. Here, we're just trying to establish CASE session.

Also, prior to getting to this line, we actually update the state here: https://github.com/project-chip/connectedhomeip/blob/32497e85cd2ad34a417835102c01ba710413bbe3/src/app/clusters/ota-requestor/OTARequestor.cpp#L476 which I also don't feel is necessary unless there's some state checking you're doing elsewhere that I'm missing.

And then again, it'll get updated once CASE session has been established here: https://github.com/project-chip/connectedhomeip/blob/32497e85cd2ad34a417835102c01ba710413bbe3/src/app/clusters/ota-requestor/OTARequestor.cpp#L377. This seems like the right place, no?

Reply:
The state must be set to something other than kIdle because in case of an error (in the same function below) we will go back to kIdle and that will be a trigger to restart the periodic timer (that's the only way we start the periodic timer). We do set the state in `TriggerImmediateQuery()` but the current function is also called from `TriggerImmediateQueryInternal()` and in does not set the state.